exercises/rotational-cipher/examples/success-text/src/RotationalCipher.hs

```diff
@@ -0,0 +1,12 @@
+module RotationalCipher (rotate) where
+
+import qualified Data.Text as T
+import           Data.Text (Text)
+import           Data.List (elemIndex)
+
+rotate :: Int -> Text -> Text
+rotate n = T.map (rotate' n ['a'..'z'] . rotate' n ['A'..'Z'])
+  where rotate' amountToRotate letters char =
+          case elemIndex char letters of
+            Nothing -> char
+            Just x  -> letters !! mod (x + amountToRotate) 26
```

Comment:
Nicely done.

I wonder, since this is a Text-based solution, if it should not demonstrate that no list traversal is necessary here.

- `rotate'` can be done purely using `isAsciiUpper`, `isAsciiLower`, `chr` and `ord` from `Data.Char`.
- `elemIndex` is quite neat for many use-cases, but since we're mainly interested in the index for later pairing with the translated `Char`, it becomes necessary to use the [partial](https://wiki.haskell.org/Avoiding_partial_functions) `!!` operator which does a linear traversal of `letters`.

Would you like to try and refine this into a solution that resembles one of the following?

```haskell
rotate :: Int -> Text -> Text
rotate n = T.map rotateLetter
  where
    rotateLetter :: Char -> Char
    rotateLetter = undefined
```

or

```haskell
rotate :: Int -> Text -> Text
rotate = T.map . rotateLetter
  where
    rotateLetter :: Int -> Char -> Char
    rotateLetter = undefined
```

That is, one that does not loop through lists of characters.

A benchmark of such a solution compared to the current one might reveal the following:

```
benchmarking rotate/x          
time                 8.388 μs   (8.385 μs .. 8.393 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 8.392 μs   (8.388 μs .. 8.398 μs)
std dev              16.58 ns   (11.70 ns .. 23.50 ns)
                               
benchmarking rotate/y          
time                 670.6 ns   (670.3 ns .. 671.0 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 670.7 ns   (670.5 ns .. 671.0 ns)
std dev              889.5 ps   (653.0 ps .. 1.367 ns)
```

for `rotate 13 "The quick brown fox jumps over the lazy dog."`.